Sources/SKCore/BuildSystemManager.swift

```diff
@@ -72,86 +131,124 @@ extension BuildSystemManager: BuildSystem {
-    if let cached: FileBuildSettings? = self.mainFileSettings[mainFile] {
-      return cached
+  /// *Must be called on queue*. Handle a request for `FileBuildSettings` on
+  /// `mainFile`. Return the new `MainFileStatus` for `mainFile`, which still
+  /// needs to be saved internally.
+  func handleFetch(
+    for mainFile: DocumentURI,
+    language: Language
+  ) -> MainFileStatus {
+    // If we already have a status for the main file, use that.
+    // Don't update any existing timeout.
+    if let status = self.mainFileStatuses[mainFile] {
+      return status
+    }
+    // This is a new `mainFile` that we need to handle. We need to fetch the
+    // build settings.
+    if let buildSystem = self.buildSystem {
+      buildSystem.registerForChangeNotifications(for: mainFile, language: language)
+
+      // Register the timeout if it's applicable.
+      if let fallback = self.fallbackBuildSystem {
+        self.queue.asyncAfter(deadline: DispatchTime.now() + self.fallbackSettingsTimeout) { [weak self] in
+          guard let self = self else { return }
+          self.handleFallbackTimer(for: mainFile, language: language, fallback)
+        }
+      }
+      return .waiting
+    } else if let fallback = self.fallbackBuildSystem {
+      // Only have a fallback build system. We consider it be a primary build
+      // system that functions synchronously.
+      if let settings = fallback.settings(for: mainFile, language) {
+        return .fallback(settings)
+      } else {
+        return .unsupported
+      }
+    } else {  // Don't have any build systems.
+      return .unsupported
     }
-    self.buildSystem?.registerForChangeNotifications(for: mainFile, language: language)
-    let settings = self.querySettings(for: mainFile, language)
-    self.mainFileSettings[mainFile] = .some(settings)
-    return settings
   }
 
-  /// *Must be called on queue*. Query the build system and/or fallback build system
-  /// for build settings.
-  private func querySettings(for file: DocumentURI, _ language: Language) -> FileBuildSettings? {
-    if let buildSystem = self.buildSystem,
-       let settings = buildSystem.settings(for: file, language) {
-      return settings
+  /// *Must be called on queue*. Update and notify our delegate for the given
+  /// main file changes if they are convertable into `FileBuildSettingsChange`.
+  func updateAndNotifyStatuses(changes: [DocumentURI: MainFileStatus]) {
+    var changedWatchedFiles = [DocumentURI: FileBuildSettingsChange]()
+    for (mainFile, status) in changes {
+      let watches = self.watchedFiles.filter { $1.mainFile == mainFile }
+      guard !watches.isEmpty else {
+        // Possible notification after the file was unregistered. Ignore.
+        continue
+      }
+      self.mainFileStatuses[mainFile] = status
+      if let change = status.buildSettingsChange {
+        for watch in watches {
+          changedWatchedFiles[watch.key] = change
```

Comment:
If the status changes but the settings are the same, we should avoid an unnecessary notification.  For example, going from waitingFallback to fallback with the same settings.

Reply:
Good call, added that but we need to be sure to emit one if we were waiting